doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,237 @@
+rather several stores may be located between the two. These will include stores to pending push slots, to determine the stack at the transition point (as these are used for rematerialization),
+and stores of call or allocation results. The fall through path will point to the inlined code,
+but the taken path will point to **OSR induce block**. OSR induce block will contain call to the **jitInduceOSRAtCurrentPC** helper and will have exception edges to the **OSR catch block**.
+
+As an optimization, we might have a store of the call result into an auto other than a pending push temp is that call results are often immediately stored to autos in the bytecode. Since we need to store it to the auto anyway, we want to avoid storing to a pending push temp, then potentially inducing OSR, and then storing again to the original auto. Instead we notice that the call result is stored into an auto (and presumably that what we store into the auto is always the result of the call, i.e. there's no branch target in between), and we push the OSR point forward past the store.
+
+Runtime assumptions are added to patch the inserted OSR guards in the event of a method redefinition,
+triggering the transition of execution to the JVM via the OSR mechanism. The OSR guards are permanently patched during an OSR yield to the JVM at which invalidation happens( e.g. a redefinition occurs). An important distinction is that all OSR guards will be patched in the event of any redefinition,
+rather than just the required set to protect a removed HCR guard. This is due to the high overhead of calculating these sets for individual HCR guards.
+
+## OSR and virtual guard merging
+
+As an optimization, we can mark some existing virtual guard as an OSR guard, therefore merging them together. This is achieved by making virtual guard kill fear. The disadvantage of the guard  merging is that the merged guard cannot be removed anymore, while the original virtual guard could've been removed as a result of later optimizaitons.
+
+
+## OSR induce blocks
+
+Since OSR catch blocks apply for any possible transition points in a method, the OSR induce block allows bookkeeping to be specified at an individual OSR point.
+It contains **irrelevant stores**, which store default values to symrefs.
+These ensure only those symrefs that are live at an OSR point appear so to other optimizations.
+Moreover, the block potentially contains rematerializaion of live symrefs and monitor enters, to further reduce the constraints on other optimizations.
+These blocks end with an OSR induce, which transitions to the VM, and an exception edge to the OSR catch block.
+Once the OSR induce is called, we transition to the VM. The VM will then create **OSRBuffer**,
+based on the size of the stack and the number of autos it expects at the induce's bytecode index.
```

Comment:
It might be due the confusing formatting. This sub-section is under the nextGenHCR section and it's mentioned at the beginning that it's a voluntary OSR. But I am not sure how to make this formatting more clear.

Reply:
except for relying on the header's fond size.

Reply:
I will check the helpers...

Reply:
Changed to:
"## OSR induce blocks

The main purpose of the OSR induce blocks is to voluntarily induce OSR.
In addition, since OSR catch blocks apply for any possible transition points in a method, the OSR induce block allow bookkeeping to be specified at an individual OSR point.
It contains **irrelevant stores**, which store default values to symrefs.
These ensure that variables that are dead at a specific OSR point don't spuriously appear to be live based on uses in the OSR code block.
Moreover, the block potentially contains rematerializaion of live symrefs and monitor enters, to further reduce the constraints on other optimizations.
The OSR induce blocks end with an OSR induce (a call to jitInduceOSRAtCurrentPC helper) which transitions to the VM, and an exception edge to the OSR catch block.
Once the induce helper is called, we transition to the VM."